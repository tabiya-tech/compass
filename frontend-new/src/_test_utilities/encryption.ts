import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
