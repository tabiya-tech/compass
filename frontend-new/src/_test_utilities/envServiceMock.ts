jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
  getApplicationLoginCode: jest.fn(() => ""),
  getApplicationRegistrationCode: jest.fn(() => ""),
  getLoginCodeDisabled: jest.fn(() => "false"),
  getRegistrationDisabled: jest.fn(() => "false"),
  getMetricsEnabled: jest.fn(() => "true"),
  getMetricsConfig: jest.fn(() => ""),
  getCvUploadEnabled: jest.fn(() => "true"),
  getSocialAuthDisabled: jest.fn(() => "false"),
}));
