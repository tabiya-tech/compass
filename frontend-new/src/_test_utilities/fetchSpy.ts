import { ExtendedRequestInit } from "src/utils/customFetch/customFetch";

export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}

// spy on authenticated fetch requests
export function setupAPIServiceSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });

  jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
  return jest.spyOn(require("src/utils/customFetch/customFetch"), "customFetch");
}

export function expectCorrectFetchRequest(fetchSpy: jest.SpyInstance, expectedUrl: string, expectedConfig: ExtendedRequestInit) {
  expect(fetchSpy).toHaveBeenCalledTimes(1);
  const [actualUrl, actualConfig] = fetchSpy.mock.calls[0];
  expect(actualUrl).toBe(expectedUrl);
  expect(actualConfig.method).toBe(expectedConfig.method);

  // Compare headers if they exist
  if(actualConfig.headers) {
    // Handle headers comparison - convert Headers object to plain object if needed
    const actualHeaders = actualConfig.headers instanceof Headers
      ? Object.fromEntries(actualConfig.headers.entries())
      : actualConfig.headers;
    expect(actualHeaders).toEqual(
      expect.objectContaining(expectedConfig.headers),
    );
  }

  // Compare bodies by parsing JSON if they exist
  if (actualConfig.body && expectedConfig.body) {
    let actualBody: unknown;
    let expectedBody: unknown;
    
    try {
      actualBody = JSON.parse(JSON.stringify(actualConfig.body));
      expectedBody = JSON.parse(JSON.stringify(expectedConfig.body));
    } catch (e) {
      console.error('Error parsing request bodies:', e);
      throw new Error('Failed to parse request bodies for comparison');
    }
    expect(actualBody).toEqual(expectedBody);
  }

  // Check any extra properties if provided
  type ExtendedConfigKey = keyof ExtendedRequestInit;
  const extraProperties: ExtendedConfigKey[] = [
    'expectedStatusCode',
    'serviceName',
    'serviceFunction',
    'failureMessage',
    'expectedContentType'
  ] as ExtendedConfigKey[];

  extraProperties.forEach((key) => {
    if (expectedConfig[key] !== undefined) {
      expect(actualConfig[key]).toEqual(expectedConfig[key]);
    }
  });
}

