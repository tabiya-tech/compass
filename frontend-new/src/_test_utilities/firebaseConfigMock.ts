import { AuthProviderIds } from "src/auth/auth.types";

const auth = jest.fn(() => ({
  signInWithCustomToken: jest.fn(),
  signInWithEmailAndPassword: jest.fn(),
  createUserWithEmailAndPassword: jest.fn(),
  signOut: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
}));

export function mockFirebaseConfig() {
  jest.spyOn(require("src/auth/firebaseConfig"), "auth").mockImplementation(auth);
}
