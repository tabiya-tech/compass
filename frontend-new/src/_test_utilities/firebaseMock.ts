import { AuthProviderIds } from "src/auth/auth.types";

const auth = {
  currentUser: {
    getIdToken: jest.fn(),
    getTokenResult: jest.fn(),
  },
  onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
  signInWithPopup: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
};

const firebase = {
  initializeApp: jest.fn(),
  auth: () => auth,
};

export default firebase;
