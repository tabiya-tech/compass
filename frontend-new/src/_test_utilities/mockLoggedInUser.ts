import { AuthContextValue, TabiyaUser, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUser;
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    updateUser: (user: TabiyaUser) => null,
    updateUserByAccessToken: () => {},
    ...mockedValues,
  });
}

export const TestUser = {
  id: "0001",
  name: "Test User",
  email: "test@email.com",
};
