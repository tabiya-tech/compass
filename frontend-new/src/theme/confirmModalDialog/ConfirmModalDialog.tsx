import React from "react";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";
import { Dialog, DialogActions, DialogContent, DialogTitle, useMediaQuery, useTheme } from "@mui/material";
import { Theme } from "@mui/material/styles";
import Typography from "@mui/material/Typography";
import SecondaryButton from "src/theme/SecondaryButton/SecondaryButton";
import CloseIcon from "@mui/icons-material/Close";
import PrimaryIconButton from "src/theme/PrimaryIconButton/PrimaryIconButton";

export interface ConfirmModalDialogProps {
  title: string;
  content: React.ReactElement;
  isOpen: boolean;
  onCancel: () => void;
  onConfirm: () => void;
  onDismiss: () => void;
  cancelButtonText: string;
  confirmButtonText: string;
  showCloseIcon?: boolean;
}

const uniqueId = "edc0ac15-67a7-4c1f-9934-94fa65757046";

export const DATA_TEST_ID = {
  CONFIRM_MODAL: `confirm-modal-${uniqueId}`,
  CONFIRM_MODAL_TITLE: `confirm-modal-title-${uniqueId}`,
  CONFIRM_MODAL_CONTENT: `confirm-modal-content-${uniqueId}`,
  CONFIRM_MODAL_CANCEL: `confirm-modal-cancel-${uniqueId}`,
  CONFIRM_MODAL_CONFIRM: `confirm-modal-confirm-${uniqueId}`,
  CONFIRM_MODAL_CLOSE: `confirm-modal-close-${uniqueId}`,
};

const ConfirmModalDialog: React.FC<ConfirmModalDialogProps> = (props) => {
  const theme = useTheme();
  const isSmallMobile = useMediaQuery((theme: Theme) => theme.breakpoints.down("sm"));

  return (
    <Dialog
      open={props.isOpen}
      onClose={props.onDismiss}
      aria-labelledby="confirm-modal"
      data-testid={DATA_TEST_ID.CONFIRM_MODAL}
      PaperProps={{
        sx: {
          display: "flex",
          gap: isSmallMobile ? theme.tabiyaSpacing.xl : theme.tabiyaSpacing.lg,
          padding: isSmallMobile
            ? theme.fixedSpacing(theme.tabiyaSpacing.md)
            : theme.fixedSpacing(theme.tabiyaSpacing.lg),
        },
      }}
    >
      <DialogTitle
        id="confirm-modal-title"
        data-testid={DATA_TEST_ID.CONFIRM_MODAL_TITLE}
        sx={{
          padding: 0,
          display: "flex",
          justifyContent: "space-between",
          alignItems: "center",
        }}
      >
        <Typography variant="h6" component="span">
          {props.title}
        </Typography>
        {props.showCloseIcon && (
          <PrimaryIconButton
            title="close confirm modal"
            onClick={props.onDismiss}
            data-testid={DATA_TEST_ID.CONFIRM_MODAL_CLOSE}
            sx={{ color: theme.palette.common.black }}
          >
            <CloseIcon />
          </PrimaryIconButton>
        )}
      </DialogTitle>
      <DialogContent data-testid={DATA_TEST_ID.CONFIRM_MODAL_CONTENT} sx={{ padding: 0 }}>
        {props.content}
      </DialogContent>
      <DialogActions sx={{ padding: 0 }}>
        <SecondaryButton onClick={props.onCancel} data-testid={DATA_TEST_ID.CONFIRM_MODAL_CANCEL}>
          {props.cancelButtonText}
        </SecondaryButton>
        <PrimaryButton onClick={props.onConfirm} data-testid={DATA_TEST_ID.CONFIRM_MODAL_CONFIRM}>
          {props.confirmButtonText}
        </PrimaryButton>
      </DialogActions>
    </Dialog>
  );
};

export default ConfirmModalDialog;
