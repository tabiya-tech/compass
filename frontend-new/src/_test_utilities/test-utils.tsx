// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
