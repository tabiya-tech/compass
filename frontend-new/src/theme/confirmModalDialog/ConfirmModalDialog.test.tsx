// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import ConfirmModalDialog, { DATA_TEST_ID } from "src/theme/confirmModalDialog/ConfirmModalDialog";

describe("ConfirmModalDialog", () => {
  test("should render component correctly", () => {
    // GIVEN the ConfirmModalDialog
    const givenNewConversationDialog = (
      <ConfirmModalDialog
        title="Sample Title?"
        content={
          <>
            This is a sample body text for the ConfirmModalDialog component.
            <br />
            <br />
            Please confirm your action.
          </>
        }
        isOpen={true}
        onCancel={() => {}}
        onConfirm={() => {}}
        cancelButtonText="Cancel"
        confirmButtonText="Confirm"
      />
    );

    // WHEN the ConfirmModalDialog is rendered
    render(givenNewConversationDialog);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND dialog container to be in the document
    const dialogContainer = screen.getByTestId(DATA_TEST_ID.CONFIRM_MODAL);
    expect(dialogContainer).toBeInTheDocument();
    // AND dialog title to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CONFIRM_MODAL_TITLE)).toBeInTheDocument();
    // AND dialog content to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CONFIRM_MODAL_CONTENT)).toBeInTheDocument();
    // AND dialog cancel button to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CONFIRM_MODAL_CANCEL)).toBeInTheDocument();
    // AND dialog confirm button to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CONFIRM_MODAL_CONFIRM)).toBeInTheDocument();
    // AND dialog to match the snapshot
    expect(dialogContainer).toMatchSnapshot();
  });
});
