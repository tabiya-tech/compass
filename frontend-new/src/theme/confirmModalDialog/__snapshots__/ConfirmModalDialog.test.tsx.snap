// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModalDialog should render component correctly 1`] = `
<div
  class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
  data-testid="confirm-modal-edc0ac15-67a7-4c1f-9934-94fa65757046"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
    role="presentation"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    tabindex="-1"
  >
    <div
      aria-labelledby="confirm-modal"
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-3kkgai-MuiPaper-root-MuiDialog-paper"
      role="dialog"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 MuiDialogTitle-root css-f196o7-MuiTypography-root-MuiDialogTitle-root"
        data-testid="confirm-modal-title-edc0ac15-67a7-4c1f-9934-94fa65757046"
        id="confirm-modal-title"
      >
        <span
          class="MuiTypography-root MuiTypography-h6 css-1o3m3rb-MuiTypography-root"
        >
          Sample Title?
        </span>
      </h2>
      <div
        class="MuiDialogContent-root css-8jfn0p-MuiDialogContent-root"
        data-testid="confirm-modal-content-edc0ac15-67a7-4c1f-9934-94fa65757046"
      >
        This is a sample body text for the ConfirmModalDialog component.
        <br />
        <br />
        Please confirm your action.
      </div>
      <div
        class="MuiDialogActions-root MuiDialogActions-spacing css-9xw8iw-MuiDialogActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-disableElevation css-187exs5-MuiButtonBase-root-MuiButton-root"
          data-testid="confirm-modal-cancel-edc0ac15-67a7-4c1f-9934-94fa65757046"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation css-fph7a5-MuiButtonBase-root-MuiButton-root"
          data-testid="confirm-modal-confirm-edc0ac15-67a7-4c1f-9934-94fa65757046"
          tabindex="0"
          type="button"
        >
          Confirm
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
